### PR TITLE
Production images and basic k8s configuration

### PR DESCRIPTION
**Pull request by cableman**

### CODE REVIEW

File: infrastructure/k8s/elasticsearch.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+              - ReadWriteOnce
+            resources:
+              requests:
+                storage: 30Gi
+            storageClassName: managed-premium-retain
+      config:
+        node.master: true
+        node.data: true
+        node.ingest: true
+        #node.store.allow_mmap: false
```
Comment (tuj) [inline infrastructure/k8s/elasticsearch.yaml]: Remove?

File: infrastructure/k8s/elasticsearch.yaml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      podTemplate:
+        spec:
+          # this changes the kernel setting on the node to allow ES to use mmap
```
Comment (tuj) [inline infrastructure/k8s/elasticsearch.yaml]: Remove this comment?

File: infrastructure/k8s/letsencrypt-clusterissuer.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+apiVersion: cert-manager.io/v1alpha2
+kind: ClusterIssuer
+metadata:
+  namespace: cover-service
+  name: letsencrypt
+spec:
+  acme:
+    server: https://acme-v02.api.letsencrypt.org/directory # production
+    #server: https://acme-staging-v02.api.letsencrypt.org/directory # staging
```
Comment (tuj) [inline infrastructure/k8s/letsencrypt-clusterissuer.yaml]: Remove?

File: infrastructure/k8s/readme.md [hunk 4]
```diff
@@ -0,0 +1,290 @@
+# Azure kubernetes service
+This document explains how to configure AKS to run Cover Service in Kubernetes with virtual node scaling and monitoring.
+
+Login in through CLI and list available regions. You should use a region in the EU to ensure data safety.
```
Comment (tuj) [inline infrastructure/k8s/readme.md]: Remove double in
"Log in through..."

File: infrastructure/k8s/readme.md [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+ksname=ddb-cover-service
+res=CoverService 
+region=westeurope
+version=$(az aks get-versions -l ${region} --query 'orchestrators[-1].orchestratorVersion' -o tsv)
+```
+
+## Create services and the cluster
+This setup is the more advanced AKS setup with virtual node 
+(https://docs.microsoft.com/en-us/azure/aks/virtual-nodes-cli) to handle peak traffic better. This requires an AKS setup
+ with advanced network setup to enabled communication between ACI (Azure container instance) and the cluster.
```
Comment (tuj) [inline infrastructure/k8s/readme.md]: Spelling
"enable"

File: infrastructure/k8s/readme.md [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+Verify that you are connected to the cluster now.
+```sh
+kubectl get nodes
+```
+
+## Helm
+We are going to use https://helm.sh/ to install ingress and cert-manager into the cluster setup. Note that we here are using helm version 3. We also install the kubectx helper tool as it makes switching cluster and namespaces easier.
+```sh
+brew install helm
```
Comment (tuj) [inline infrastructure/k8s/readme.md]: Only mac ?

File: infrastructure/k8s/readme.md [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+```
+
+Next enable the monitoring add-on by running the command below.
+```sh
+az aks enable-addons --resource-group ${res} \
+    --name ${ksname} \
+    --addons monitoring \
+    --workspace-resource-id ${workspaceresoid}
+```
+
```
Comment (tuj) [inline infrastructure/k8s/readme.md]: Remove extra empty line

File: infrastructure/k8s/readme.md [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+  apiGroup: rbac.authorization.k8s.io
+```
+
+```sh
+kubectl apply -f logreader-rbac.yaml
+```
+
+## Horizontal pod autoscaler (HPA)
+
+The application deployment use HPA see the app deployment for more information.
```
Comment (tuj) [inline infrastructure/k8s/readme.md]: uses

Comment (tuj) [reply infrastructure/k8s/readme.md]: "The application deployment uses HPA. See the app deployment for more information."

File: infrastructure/k8s/readme.md [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+
+# Install elastic search operator
+
+```sh
+kubectl apply -f https://download.elastic.co/downloads/eck/1.0.0-beta1/all-in-one.yaml
+```
+
+Use this after you have installed the elastic-search application.
```
Comment (tuj) [inline infrastructure/k8s/readme.md]: A bit cryptic comment...

File: infrastructure/k8s/readme.md [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+
+# Github package registry
+To use the github package registry from kubernetes you need to store basic authentication in a secret and use it in `imagePullSecrets` in the yaml deployment.
+
+```sh
+kubectl create secret docker-registry github-registry \
+    --docker-server=docker.pkg.github.com \
+    --docker-username=cableman \
+    --docker-password=XXXXX_TOKEN_XXXXX \
+    --docker-email=jeskr@aarhus.dk
```
Comment (tuj) [inline infrastructure/k8s/readme.md]: Remove mail?

File: infrastructure/k8s/readme.md [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+```
+
+See the github secret in clear text from the cluster.
+```sh
+kubectl get secret github-registry --output="jsonpath={.data.\.dockerconfigjson}" | base64 --decode
+```
+
+# Application install
```
Comment (tuj) [inline infrastructure/k8s/readme.md]: Missing comments about what is going on...

File: infrastructure/docker/cover-service-jobs/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+COPY --from=docker.pkg.github.com/danskernesdigitalebibliotek/ddb-cover-service/cover-service ${APP_PATH}/ ${APP_PATH}/
+
+# Install configuration template handler
+ADD https://github.com/kelseyhightower/confd/releases/download/v0.16.0/confd-0.16.0-linux-amd64 /usr/local/bin/confd
+RUN chmod +x /usr/local/bin/confd
+
+COPY docker-entrypoint.sh /usr/local/bin/
+RUN chmod +x /usr/local/bin/docker-entrypoint.sh
+
+CMD [ "docker-entrypoint.sh" ]
```
[the remark this answers is not in the document]
Comment (cableman) [reply infrastructure/docker/cover-service-jobs/Dockerfile]: Fixed

[the remark this answers is not in the document]
Comment (cableman) [reply infrastructure/docker/cover-service-jobs/Dockerfile]: Fixed

File: infrastructure/docker/cover-service-jobs/etc/confd/conf.d/env.local.toml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+[template]
+src = "env.local.tmpl"
+dest = "/var/www/html/.env.local"
+mode = "0644"
+keys = [
+    "/app-config"
+]
```
[the remark this answers is not in the document]
Comment (cableman) [reply infrastructure/docker/cover-service-jobs/etc/confd/conf.d/env.local.toml]: Fixed

File: infrastructure/docker/cover-service-jobs/etc/supervisor.d/background_queue.ini [hunk 14]
```diff
@@ -0,0 +1,11 @@
+[program:background_qeueu]
```
[the remark this answers is not in the document]
Comment (cableman) [reply infrastructure/docker/cover-service-jobs/etc/supervisor.d/background_queue.ini]: Fixed

File: infrastructure/docker/cover-service-jobs/etc/supervisor.d/search_queue.ini [hunk 15]
```diff
@@ -0,0 +1,11 @@
+[program:search_qeueu]
```
[the remark this answers is not in the document]
Comment (cableman) [reply infrastructure/docker/cover-service-jobs/etc/supervisor.d/search_queue.ini]: Fixed

File: infrastructure/docker/php7.2-fpm/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        pdo_mysql \
+        soap \
+        sysvsem
+
+RUN apk --update add --no-cache --virtual .build-deps autoconf g++ make \
+    && pecl install redis-4.0.1 imagick \
+    && docker-php-ext-enable redis imagick \
+    && apk del .build-deps
+
+# Added FPM healtcheck script (https://github.com/renatomefi/php-fpm-healthcheck)
```
[the remark this answers is not in the document]
Comment (cableman) [reply infrastructure/docker/php7.2-fpm/Dockerfile]: Fixed

File: infrastructure/docker/cover-service-jobs/etc/supervisor.d/cover_queue.ini [hunk 17]
```diff
@@ -0,0 +1,11 @@
+[program:cover_qeueu]
```
[the remark this answers is not in the document]
Comment (cableman) [reply infrastructure/docker/cover-service-jobs/etc/supervisor.d/cover_queue.ini]: Fixed

File: infrastructure/k8s/elasticsearch.yaml (hunk 1 above)
Comment (cableman) [reply infrastructure/k8s/elasticsearch.yaml]: Fixed

File: infrastructure/k8s/app-deployment.yaml [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+          exec:
+            command:
+              - php-fpm-healthcheck
+          initialDelaySeconds: 1
+          periodSeconds: 5
+        livenessProbe:
+          exec:
+            command:
+              - php-fpm-healthcheck
+              #- --accepted-conn=50000
```
[the remark this answers is not in the document]
Comment (cableman) [reply infrastructure/k8s/app-deployment.yaml]: Fixed

File: infrastructure/k8s/letsencrypt-clusterissuer.yaml (hunk 3 above)
Comment (cableman) [reply infrastructure/k8s/letsencrypt-clusterissuer.yaml]: Nope, it's there to ensure that we can remeber the stg url for the stg cluster

File: infrastructure/k8s/elasticsearch.yaml (hunk 2 above)
Comment (cableman) [reply infrastructure/k8s/elasticsearch.yaml]: Updated.

File: infrastructure/k8s/readme.md (hunk 4 above)
Comment (cableman) [reply infrastructure/k8s/readme.md]: Fixed

File: infrastructure/k8s/readme.md (hunk 5 above)
Comment (cableman) [reply infrastructure/k8s/readme.md]: Fixed

File: infrastructure/k8s/readme.md (hunk 6 above)
Comment (cableman) [reply infrastructure/k8s/readme.md]: Yes, as i don't know how to install is in Windows or Linux.

File: infrastructure/k8s/readme.md (hunk 7 above)
Comment (cableman) [reply infrastructure/k8s/readme.md]: Fixed

File: infrastructure/k8s/readme.md (hunk 8 above)
Comment (cableman) [reply infrastructure/k8s/readme.md]: Fixed

File: infrastructure/k8s/readme.md (hunk 9 above)
Comment (cableman) [reply infrastructure/k8s/readme.md]: Fixed

File: infrastructure/k8s/readme.md (hunk 10 above)
Comment (cableman) [reply infrastructure/k8s/readme.md]: Fixed

File: infrastructure/k8s/readme.md (hunk 11 above)
Comment (cableman) [reply infrastructure/k8s/readme.md]: Fixed

File: infrastructure/k8s/readme.md [hunk 19]
```diff
@@ -259,7 +259,7 @@ kubectl create secret docker-registry github-registry \
     --docker-server=docker.pkg.github.com \
     --docker-username=cableman \
```
Comment (tuj) [inline infrastructure/k8s/readme.md]: YOUR_USERNAME

**Pull request merged**